OWL2/Medusa.hs

```diff
@@ -0,0 +1,131 @@
+
+import Data.Maybe
+import qualified Data.Set as Set
+
+data Medusa = Medusa {
+               indivs :: Set.Set (QName,QName),
+               relations :: Set.Set (QName, QName, QName, QName)}
+
+-- | given an OWL ontology (iri and theory), compute the medusa data
+medusa :: IRI.IRI -> (Sign, [Named Axiom])
+                       -> Result Medusa
+medusa _ (sig, nsens) = do
+  let inds = individuals sig
+      getC = getClass (map sentence nsens)
+      getR tInds = getMeetsFacts (map sentence nsens) tInds
+      allInds = Set.map (\ i -> (i,getC i)) inds
+      relTuples = foldl Set.union Set.empty $
+                  map (getR allInds) $ Set.toList inds
+      images = Set.foldl Set.union Set.empty $
+               Set.map (\(i1, _, i2, _) -> Set.fromList [i1, i2]) relTuples
+  return $ Medusa {
+            indivs = Set.filter (\(i,_) -> Set.member i images) allInds ,
+            relations = relTuples
+           }
+
+-- | get the class of an individual
+getClass :: [Axiom] -> QName -> QName
+getClass axs n = case mapMaybe (getClassAux n) axs of
+   (c:_) -> c
+   [] -> nullQName { localPart = "unknown" }
+
+getClassAux :: QName -> Axiom -> Maybe QName
+getClassAux ind ax =
+  case axiomTopic ax of
+    SimpleEntity e | cutIRI e == ind ->
+      case axiomBit ax of
+         ListFrameBit (Just Types) (ExpressionBit classes) -> firstClass classes
+         _ -> Nothing
+    _ -> Nothing
+
+--  for each individual "p1" that has a fact "meets p2"
+--  look for individuals "i1" and "i2" such that
+--  i1 has_fiat_boundary p1 and i2 has_fiat_boundary p2
+--  and return i1 type(p1) i2 type(p2)
+getMeetsFacts :: [Axiom] -> Set.Set (QName, QName) -> QName ->
+              Set.Set (QName, QName, QName, QName)
+getMeetsFacts axs tInds n = case mapMaybe (getMeetsFactsAux axs tInds n) axs of
+   [] -> Set.empty
+   x -> Set.fromList x
+
+getMeetsFactsAux :: [Axiom] -> Set.Set (QName, QName) -> QName -> Axiom -> Maybe (QName, QName, QName, QName)
+getMeetsFactsAux axs tInds point1 ax =
+  case axiomTopic ax of
+    SimpleEntity e | cutIRI e == point1 ->
+      case axiomBit ax of
+         ListFrameBit Nothing (IndividualFacts [([],
+                               (ObjectPropertyFact Positive
+                                                   (ObjectProp ope) point2))
+                                               ]) ->
+            if localPart ope == "meets" then
+                getFiatBoundaryFacts axs tInds point1 point2
+              else Nothing
+         _ -> Nothing
+    _ -> Nothing
+
+getFiatBoundaryFacts :: [Axiom] -> Set.Set (QName, QName) -> QName -> QName ->
+                     Maybe (QName, QName, QName, QName)
+getFiatBoundaryFacts axs tInds point1 point2 =
+   let i1 = case mapMaybe (getFiatBoundaryFactsAux point1) axs of
+              (c:_) -> Just c
+              [] -> Nothing
+       i2 = case mapMaybe (getFiatBoundaryFactsAux point2) axs of
+              (c:_) -> Just c
+              [] -> Nothing
```

Comment:
this pattern occurs 3 times (also in function `getClass`, if you add a call of function `maybe` there) and should be relegated to an own function.